### PR TITLE
change the while of get_results + add time duration of jobs

### PR DESCRIPTION
**Pull request by jplassmann**

### CODE REVIEW

File: python-lib/amazon_transcribe_api_client.py [hunk 1]
```diff
@@ -198,6 +198,15 @@ def get_list_jobs(self,
             i += 1
             if len(response.get("TranscriptionJobSummaries", [])) == 0 or next_token is None:
                 break
+
+        for job in result:
```
[the remark this answers is not in the document]
Comment (jplassmann) [reply python-lib/amazon_transcribe_api_client.py]: done in 23a19569b6de334b5677657c8e47cdbd7123949e

File: python-lib/amazon_transcribe_api_client.py [hunk 2]
```diff
@@ -318,3 +321,25 @@ def _result_parser(self,
...
+                            job_res_data: Dict):
+        """
+        If the job duration is longer than the timeout setup previously, a JOB_TIMEOUT_ERROR will
+        be written in the column error of the Dataframe, otherwise it returns None.
+        """
+
+        date_job_created = job_summary.get("CreationTime")
+        now = datetime.datetime.now(tz=date_job_created.tzinfo)
+        time_delta_sec = (now - date_job_created).seconds
+        time_delta_min = time_delta_sec / 60
```
Comment (MayeulRousselet) [inline python-lib/amazon_transcribe_api_client.py:336]: no need to store seconds here, directly assign to minutes

**Review by MayeulRousselet — CHANGES_REQUESTED** [review]
I am wondering if we should call Transrbie api to cancel a job which reach the timeout on our side or not. WDYT?

File: custom-recipes/amazon-transcribe/recipe.py [hunk 3]
```diff
@@ -9,7 +10,7 @@
 # CONSTANT DEFINITION
 # ==============================================================================
 
-RECIPE_JOB_ID = dataiku.dku_custom_variables.get('jobId')
+RECIPE_JOB_ID = f"{dataiku.dku_custom_variables.get('jobId')}_{uuid.uuid4().hex}"
```
[the remark this answers is not in the document]
Comment (jplassmann) [reply custom-recipes/amazon-transcribe/recipe.py]: done in 88b96bbe1935d7840e46260a3c3ddc5be2ad7b0c

File: python-lib/amazon_transcribe_api_client.py [hunk 4]
```diff
@@ -318,3 +321,25 @@ def _result_parser(self,
             logging.warning(f"Unknown state encountered: {job_status}")
             raise UnknownStatusError(f"Unknown state encountered: {job_status}")
         return job_data
+
+    @staticmethod
+    def handle_job_duration(job: Dict,
```
[the remark this answers is not in the document]
Comment (jplassmann) [reply python-lib/amazon_transcribe_api_client.py]: done in 88b96bbe1935d7840e46260a3c3ddc5be2ad7b0c

File: python-lib/amazon_transcribe_api_client.py (hunk 2 above)
Comment (jplassmann) [reply python-lib/amazon_transcribe_api_client.py]: done in 88b96bbe1935d7840e46260a3c3ddc5be2ad7b0c

File: custom-recipes/amazon-transcribe/recipe.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jplassmann) [reply custom-recipes/amazon-transcribe/recipe.py]: done better in 8baf802fcfa96a460727cd2c201d0365089b56cb

File: custom-recipes/amazon-transcribe/recipe.json [hunk 5]
```diff
@@ -202,6 +202,23 @@
             "mandatory": true,
             "defaultValue": false
         },
+        {
+            "name": "use_timeout",
+            "label": "Use Timeout",
+            "type": "BOOLEAN",
+            "description": "Raise a timeout error in the error column if the job is longer than the specified timeout.",
```
Comment (andreybavt) [inline custom-recipes/amazon-transcribe/recipe.json:209]: To me the description is too technical. WDYT about: "interrupt execution if transcriptoin job takes more time than a limit"

I thought that we also discussed an option where the timeout is always set (default 2 hours). In this case I'd only keep 1 input (no checkbox needed) and put a hint that empty value means no timeout.

File: python-lib/amazon_transcribe_api_client.py [hunk 6]
```diff
@@ -318,3 +324,24 @@ def _result_parser(self,
...
+        be written in the column error of the Dataframe, otherwise it returns None.
+        """
+
+        date_job_created = job_summary.get("CreationTime")
+        now = datetime.datetime.now(tz=date_job_created.tzinfo)
+        time_delta_min = int((now - date_job_created).seconds / 60)
+        logging.info(
+            f"{job_summary.get('TranscriptionJobStatus')} | {job_summary.get('TranscriptionJobName')} | {time_delta_min} min")
+        if self.use_timeout and time_delta_min >= self.timeout:
+            logging.error(f'Job {job_summary.get("TranscriptionJobName")} timeout!')
```
Comment (andreybavt) [inline python-lib/amazon_transcribe_api_client.py:342]: This should rather be a warning than an error. I'd also change the text to be more informative: "Job {jobname} canceled after a timeout of {self.timeout} minutes"

We also need to cancel the amazon job when the timout happens

File: custom-recipes/amazon-transcribe/recipe.py [hunk 7]
```diff
@@ -1,45 +1,46 @@
...
 from plugin_params_loader import RecipeID
 
+
 # ==============================================================================
-# CONSTANT DEFINITION
+# CONSTANT/PATTERN DEFINITION
 # ==============================================================================
 
-RECIPE_JOB_ID = dataiku.dku_custom_variables.get('jobId')
+RECIPE_JOB_ID_PATTERN = f"{dataiku.dku_custom_variables.get('jobId')}_{uuid.uuid4().hex}"
```
Comment (andreybavt) [inline custom-recipes/amazon-transcribe/recipe.py:14]: nitpick, I wouldn't call it a pattern since it's an actual value

**Review by andreybavt — CHANGES_REQUESTED** [review]

File: custom-recipes/amazon-transcribe/recipe.py (hunk 7 above)
Comment (jplassmann) [reply custom-recipes/amazon-transcribe/recipe.py]: done in 1b4d2379271f3e34715ff74b42eb6f17053e741a

File: python-lib/amazon_transcribe_api_client.py (hunk 6 above)
Comment (jplassmann) [reply python-lib/amazon_transcribe_api_client.py]: done in 1b4d2379271f3e34715ff74b42eb6f17053e741a

ps: as we said before, we cannot cancel an "in progress" job

File: custom-recipes/amazon-transcribe/recipe.json (hunk 5 above)
Comment (jplassmann) [reply custom-recipes/amazon-transcribe/recipe.json]: done in e1755ff1b2fdee478584ddfaf584383c280bb6a9

File: custom-recipes/amazon-transcribe/recipe.json [hunk 8]
```diff
@@ -202,6 +202,23 @@
...
+            "mandatory": true,
+            "defaultValue": false
+        },
+        {
+            "name": "timeout",
+            "label": "timeout (min)",
+            "type": "INT",
+            "description": "Time in minutes after which the AWS job return an error if it is not completed",
```
Comment (andreybavt) [inline custom-recipes/amazon-transcribe/recipe.json]:
```suggestion
            "description": "Time in minutes after which the AWS transcription job will be abandoned resulting in an error message for a given file in the output dataset",
```

**Review by andreybavt — COMMENTED** [review]
Small suggestion regarding the timeout hint message, otherwise LGTM

File: custom-recipes/amazon-transcribe/recipe.json [hunk 9]
```diff
@@ -202,6 +202,14 @@
             "mandatory": true,
             "defaultValue": false
         },
+        {
+            "name": "timeout_min",
+            "label": "timeout (min)",
+            "type": "INT",
+            "description": "Time in minutes after which the AWS job return an error if it is not completed. Hint: leave it empty if no timeout is wanted.",
```
Comment (andreybavt) [inline custom-recipes/amazon-transcribe/recipe.json:209]:
```suggestion
            "description": "Time in minutes after which the AWS transcription job will be abandoned resulting in an error message for a given file in the output dataset. Leave empty for unlimited waiting time.",
```

**Review by andreybavt — APPROVED** [review]

**Review by MayeulRousselet — APPROVED** [review]

**Pull request merged**